### ISSUE TITLE
om bosh-env / get-credentials / etc. feedback

### ISSUE THREAD

**Issue by aegershman** [pivotal-cf/om#385]
related: #290 

As admins, in order to more-efficiently manage credentials and bosh targets in multiple environments, we are using `eval $(om bosh-env)` more. We also are finding a lot of value in `om --env my-env.ml credentials <xyz>` Here is some feedback. 

I can break these out into separate issues if preferred?

## ability to logout (unset) env vars

`bosh-env` exports helpful `env` vars. But without `unset`'ing them, you may inadvertently target the wrong foundation or credhub. Unsetting the env vars require clearing the shell with the following:

```sh
unset BOSH_CLIENT
unset BOSH_CLIENT_SECRET
unset CREDHUB_CLIENT
unset CREDHUB_SECRET
unset CREDHUB_SERVER
unset CREDHUB_CA_CERT
unset BOSH_CA_CERT
unset BOSH_ENVIRONMENT
unset BOSH_ALL_PROXY
unset CREDHUB_PROXY
```

It would be valuable for `om` to have a `logout` analogy which unsets any of the env vars it leverages, so that `eval $(om bosh-env-logout)` (or some equivalent) could give reassurance that the `env` is clean.

Or perhaps simply `om logout`, which unsets any/all env vars, not just the `CREDHUB_*` and `BOSH_*` vars? Or it could be made more specific with `om logout --bosh --credhub`, etc.?

## selectively export `BOSH_*` keys, not `CREDHUB_*`

Some of our automation tasks use `om bosh-env` to get the director credentials, but in the same script will use a call to a *different* credhub than the bosh director's credhub... But `om bosh-env` will have the environment vars targeting the _director's_ credhub.

It would be valuable to be able to specify "only export `BOSH_*`" params, so that the existing `CREDHUB_*` params aren't clobbered. E.g., 

```sh
om bosh-env -e some-auth-file.yml \
  --ssh-private-key=some-private-key-used-for-socks-proxy.pem \
  --bosh-only

# or --credhub-only?
```

... or something to that effect.

## login flow

One of the only downsides of using `om` for interfacing with `om` is having to have either basic auth user credentials or client credentials on my machine in plaintext for specifying the `--env `auth rather than being able to leverage an `--sso` workflow equivalent.

## feedback, general

I really like the `om bosh-env` command as the preferred way to target environments. It's also very valuable for extracting credentials such as a `uaa admin client` via `om -e myenv.yml credentials --product-name=cf --credential-reference=.uaa.admin-client` (or something like that)

Thanks for implementing it, and I hope it continues to get love as the way to get credentials from opsmgr, both for automation && for operators.

Thanks for your time and consideration.

**Comment by jpmcb** [comment]
Awesome! Thanks for the great feedback and suggestions:

1. Would the proposed `logout` behavior completely _clear_ the environment, or would you expect this to act like a login / logout stack? (i.e. if there was a previous environment set before using a `om bosh-env` command, this env would be restored after logout?) Or would we expect the user to be required to always run `eval $(om bosh-env)` (or some equivalent) right afterwards to set a new env?

2. We have recognized that there are some gaps in using credhub with platform automation. This appears to be another one. It makes sense that you wouldn't want the pre-exisiting `CREDHUB` target to be swallowed by the new eval.

3. I'm not sure how an SSO workflow would happen in `om`. Would this be key based? Can you clarify or show a workflow?

Thanks again for the awesome feedback. We will definitely continue to optimize the `om bosh-env`. This feedback is probably good in here for now; we will discuss this and we will get back to you if we need these in separate issues for some reason!

**Comment by kcboyle** [comment]
@aegershman apologies for the late reply on this one! could you expand some more on some of the questions above?

**Comment by aegershman** [comment]
@kcboyle no that's my bad for missing this, I'll write up a response in the next day or two 👍 standby

**Comment by aegershman** [comment]
@jpmcb sorry for delay in writing this up my bad

1. I don't think I'd expect the previous session to be restored, because the way `eval $(om print-env)` is used already assumes you're passing in the `om` creds each time: there isn't really a standalone `om login` which sets a "session". If a user runs `eval $(om bosh-env)` it exports the values into their current shell env, but it's almost always performed as `eval $(om --env sandbox-aws.yml bosh-env...)` or `eval $(om --client-id foo --client-secret bar bosh-env...)`, etc.

I don't think there is any other way to get the env vals into the current shell from a subshell except for `eval`... Including _unsetting_ env vars. So doing `om logout` (or whatever it's named) would have to be ran as `source $(om logout)`, where `logout` would allow for the env values to be cleared.

**TL;DR**, if `eval $(om bosh-env...)` is used, I'd expect `source $(om unset-bosh-env)` (or `$(om logout) or whatever would be most appropriate?)  to clear env vars and have a clean shell slate.

The only other way I could (maybe?) see this working would be is if `om` were used as a "prefix" to issue certain commands, like `om --env sandbox-aws.yml bosh -d service-instance foo xyz`, e.g. having `om` act as the auth pass-through to issue `bosh` or `credhub` params but only for the context of that request. That way when you do `om <auth method like --client-id> bosh <bosh-commands>`, the lifespan of acquiring the `bosh` creds would exist only in that one `om` call, and the user's environment would remain completely clear.

This is kinda-sorta how the [`hammer`](https://github.com/pivotal/hammer) cli works, where you can use `hammer bosh` to _show_ bosh creds (and thus could do `eval $(hammer -t sandbox bosh)` to export the values into the user shell) or pass-through the command with `hammer -t sandbox bosh <bosh commands>`, and the user's env would remain clear. The awesome thing about _that_ is then you can have multiple targets in automation or on the operator's machine, and don't have to worry (as much) about the env; you can namespace each bosh env with the `--target` or `--env` flag, similar to `fly -t` or `bosh -e`, etc. 

It's not a massive priority though, I'm just speculating.

2. Again, not a huge deal, I like the fact that it gets access to the bosh director's credhub at all since otherwise it'd be kind of a pain. It's okay that it clobbers the current `CREDHUB_*` env, it'd just be nice to have a convenience method to unset those env vars so I could log back in again. No reason to get fancy with restoring previous env values, etc. This could also benefit from using `om` as a "passthrough", e.g. `om --env sandblorp.yml credhub find -n /foo`, because then I could use the "outer" `env` to get a `credhub` value from <some-other-credhub>, and set it to the director credhub via `om` without worrying about clobbering the "outer" env, and without worrying about the env being "tainted" with the credhub targeted with `om`. Once the `om` command completes, that's it.

3. It'd have to use `https://pcf.example.io/uaa/login` to get an sso token from the currently authenticated user to allow an `om` request to go through. E.g., I'm logged in to the opsmgr console, and when I do `eval $(om --env sandbox.yml bosh-env)` or `eval $(om --target https://pcf.example.io --sso bosh-env)` or something similar, it'd halt and request a token obtained from `https://pcf.example.io/uaa/login` before giving me the commands. That way as a human user (not automation) I don't need to store client creds in a local `env.yml`, I can just rely on however I'm auth'd against the opsmgr (like SAML/LDAP/etc.)... Someone smarter than me should evaluate this workflow, though. Again, not essential, just an idea stolen from `uaa/cf/fly` CLIs.

thoughts? no rush to respond 👍

**Comment by anEXPer** [comment]
Thanks again for the suggestions, @aegershman.

Let's go ahead and break this out into separate issues. We're reviewing `om` issues and as it stands, there's too much going on in this one for us to get any traction. If you wanna break it out into the three elements and link back to this one, we'll probably be able to get somewhere on at least some of them.

We'll probably break this up ourselves, at some point, but at this very moment we're headed into a meeting, so - it's a race! And we're deploying the "slow and steady" strategy.

**Comment by kcboyle** [comment]
split these out into three issues. The first two seem reasonable nice-to-haves. The last I'm a little more iffy about, so I threw a "discussion" label on it. Let's talk about the technical implications on that one, and whether it is feasible there. 

Closing this issue. Thanks for the discussion @aegershman ! and apologies for the delay in responses!

**Comment by aegershman** [comment]
Of course, yeah and do keep in mind most of these are just brain-dumping ideas, I'm not sure if I completely agree with most of these even after writing them up. So I apologize for just plopping the ideas down. Thanks again y'all, no worries *at all* on async responses.
